### PR TITLE
Add support for .app domains

### PR DESCRIPTION
**Pull request by fideloper**
Using `generic_parser_b` to add support for Google's `.app` domains.

### CODE REVIEW

**Comment by fideloper** [comment]
I'm not sure how this testsuite works - is there an entry to `testsuite.txt` I should add for a `.app` domain?

File: src/whois.app.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+require_once('whois.parser.php');
+
+class app_handler {
+
+    function parse($data_str, $query) {
+        $r = array();
+        $r['regrinfo'] = generic_parser_b($data_str['rawdata'], array(), 'mdy');
+//        $r['regyinfo'] = array(
+//            'referrer' => 'http://www.neulevel.biz',
```
Comment (jsmitty12) [inline src/whois.app.php]: Will you set `referrer` to be `https://www.registry.google`?

File: src/whois.app.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+//            'registrar' => 'NEULEVEL'
```
Comment (jsmitty12) [inline src/whois.app.php]: Will you set `registrar` to be `Google Registry`?

**Comment by jsmitty12** [comment]
Thanks for your PR!

> I'm not sure how this testsuite works - is there an entry to testsuite.txt I should add for a .app domain?

The trick here will be finding legitimate .app domains that are likely to exist forever. Several domains were added to the test suite at one point and they no longer exist now, which causes some of the tests to fail (well, at least they're useless). I've starting using `google.tld` as a test domain assuming that Google will be around forever and that they'll register a domain on (nearly) every tld. (Ideally I would like to have two domains to test for each tld, but I've had trouble finding domains that I'm confident will still exist in 5-10 years.)

`testsuite.txt` is part of the original test suite (which I'm slowly trying to eliminate) which operates on live domains, and I don't mind if that gets updated. You need to add a new entry to the `test.txt` file in the _Other_ section... something like this:
```
za.net		sex.za.net
app		google.app
```

The new pattern I'm taking can be seen by looking in the `/tests` directory. You can use any of the existing handler tests as a pattern. Don't forget to make a fixture by capturing the whois information to a text file in the `tests/fixtures` directory so the tests will work without having to make a live query.

File: src/whois.app.php (hunk 1 above)
Comment (jsmitty12) [reply src/whois.app.php]: wait... this may not be needed. I think the code may pull this from the whois results automatically.

**Comment by fideloper** [comment]
I searched for .app domains that google has reserved, and they use [https://get.app/](https://get.app/), so that looks like a good candidate to test with.

**Comment by fideloper** [comment]
Tests added within `/tests` dir 🤟🏻

File: tests/Handlers/AppHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    protected $handler;
+
+    /**
+     * @return void
+     */
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->handler            = new \org_handler();
```
Comment (jsmitty12) [inline tests/Handlers/AppHandlerTest.php]: This is using the org_handler, not your new app_handler.

File: tests/Handlers/AppHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ *
+ * @copyright Copyright (c) 2018 Joshua Smith
+ */
+
+namespace phpWhois\Handlers;
+
+/**
+ * OrgHandlerTest
```
Comment (jsmitty12) [inline tests/Handlers/AppHandlerTest.php]: Nitpick: wrong class mention

**Comment by jsmitty12** [comment]
> I searched for .app domains that google has reserved, and they use https://get.app/, so that looks like a good candidate to test with. (And google.app does have a WHOIS entry so that should work).

I agree that `get.app` is a good candidate--its data looks a different from `google.app`, so let's do both!

I also noticed that most of the data is returned as "REDACTED FOR PRIVACY", which makes testing a little less definite if we start checking specific values. How are you searching for `.app` domains? Do you think one might exist that isn't mostly "REDACTED FOR PRIVACY"?

File: src/whois.app.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ *
+ * @link http://phpwhois.pw
+ * @copyright Copyright (C)1999,2005 easyDNS Technologies Inc. & Mark Jeftovic
+ * @copyright Maintained by David Saez
+ * @copyright Copyright (c) 2014 Dmitry Lukashin
+ */
+
+if (!defined('__APP_HANDLER__'))
```
Comment (jsmitty12) [inline src/whois.app.php]: Nitpick: add the curly braces, please

File: src/whois.app.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    define('__APP_HANDLER__', 1);
+
+require_once('whois.parser.php');
+
+class app_handler {
```
Comment (jsmitty12) [inline src/whois.app.php]: Nitpick: brace on new line

File: src/whois.app.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    function parse($data_str, $query) {
```
Comment (jsmitty12) [inline src/whois.app.php]: Nitpick: brace on new line
Nitpick: missing docblock

File: src/whois.app.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        $r = array();
+        $r['regrinfo'] = generic_parser_b($data_str['rawdata'], array(), 'mdy');
```
Comment (jsmitty12) [inline src/whois.app.php]: Nitpick: the 2nd and 3rd parameters should not be necessary--these are the defaults.

File: src/whois.app.php (hunk 5 above)
Comment (fideloper) [reply src/whois.app.php]: This one is inconsistent with if statements in every other `whois.*.php` files.

File: src/whois.app.php (hunk 6 above)
Comment (fideloper) [reply src/whois.app.php]: This one is inconsistent with other `*_handler` classes.

File: src/whois.app.php (hunk 7 above)
Comment (fideloper) [reply src/whois.app.php]: This one is inconsistent with other *_handler classes.

**Comment by fideloper** [comment]
So for some of this feedback, your suggestions (while I'm happy to do them!) are inconsistent with the rest of the code.

Let me know if you want to apply those changes anyway, I'm more than happy to!

File: src/whois.app.php (hunk 5 above)
Comment (jsmitty12) [reply src/whois.app.php]: You're right that it's inconsistent with most of the other files. I'm working on updating the code to match the PSR-1 and PSR-2 standards. I've been making the style updates to the handlers as I've created the new-style tests.

**Comment by jsmitty12** [comment]
The tests look great--thank you for this addition! I'll accept the PR once you've finished the code style updates.

**Comment by fideloper** [comment]
Think I got everything updated there - thanks for your time and feedback!

**Pull request merged**